### PR TITLE
Impaler Burrower strain (with fixes)

### PR DESCRIPTION
**Pull request by Unknownity**
# About the pull request

First of all, thank you Kugamo for making the sprites for the strain and its effects.

https://streamable.com/4m3rbr (some things are changed since this video)

A burrower strain that can deal damage while burrowed, causing enemies to have to be creative when attacking, removing its weeds will force it to surface.

Impaler abilities:
Burrowed Spikes:
Description: Sends out fast spikes in a direction that hits tile by tile.
Cost: 50 Plasma
Cooldown: 8 Seconds.
Range: 5 Tiles
Damage: 35
Can be used while burrowed.

Sunken Tail:
Description: After a windup, strikes the targeted area damaging anyone in the tile.
Cost: 100 Plasma
Cooldown: 13 seconds.
Range: 7
Damage: 60
Windup: .8 Seconds

Ensconce:
Description: After a windup, go halfway into the ground in a fortified stance (prevents you from moving) which increases your vision range by 3, increases your Burrowed spikes ability range by 2, increases your Sunken Tail ability range by 2, and damage on both abilities by 30.
Modifies Sunken Tail to do an AoE effect:
https://streamable.com/xy0tq1 (middle is full damage + ensconce'ed damage bonus, edges are normal damage)

Cost: None
Cooldown: 3 Seconds
Buffs: Armor deflection increased by 25 and armor explosive resistance increased by 50 and small explosives cannot stun the target while in the stance.

Base Tail Stab:
Can be used while burrowed.

# Explain why it's good for the game

New (actually balanced on launch) xeno content which they need.

# Testing Photographs and Procedure

<details>
<summary>Screenshots & Videos</summary>

Put screenshots and videos here with an empty line between the screenshots and the `<details>` tags.

</details>

# Changelog

:cl: Unknownity, Kugamo
add: Added new strain for the Burrower, "Impaler".
fix: Fixed burrowed mobs being able to be targeted by sentries, mines and SG autofire.
fix: Fixed burrowed mobs being able to grab mobs on the surface.
fix: Fixed burrowed mobs being able to resist while burrowed.
fix: Fixed burrowers taking damage from direct flame and shrapnel from explosions.
fix: Fixed burrowers being able to get slashed from enemy Xenos on the surface.
fix: Fixed burrowers unburrow stun to now properly target and stun enemy Xenos.
soundadd: Added sounds for base and Impaler burrower for burrowing and unburrowing, added sounds for Impaler burrower's attacks.
imageadd: Added sprites and effects for Impaler burrower by Kugamo.
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/xenomorph/mutators/strains/burrower/impaler.dm [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+		/datum/action/xeno_action/activable/corrosive_acid,
+		/datum/action/xeno_action/activable/place_construction,
+		/datum/action/xeno_action/onclick/plant_weeds,
+		/datum/action/xeno_action/onclick/place_trap
+	)
+	mutator_actions_to_add = list(
+		/datum/action/xeno_action/activable/burrowed_spikes,
+		/datum/action/xeno_action/activable/sunken_tail,
+		/datum/action/xeno_action/onclick/burrow,
+		/datum/action/xeno_action/onclick/ensconce
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/mutators/strains/burrower/impaler.dm:22]:
```suggestion
		/datum/action/xeno_action/onclick/place_trap,
	)
	mutator_actions_to_add = list(
		/datum/action/xeno_action/activable/burrowed_spikes,
		/datum/action/xeno_action/activable/sunken_tail,
		/datum/action/xeno_action/onclick/burrow,
		/datum/action/xeno_action/onclick/ensconce,
```

File: code/game/machinery/doors/airlock_types.dm [hunk 2]
```diff
@@ -827,6 +827,11 @@
 /obj/structure/machinery/door/airlock/dropship_hatch/ex_act(severity)
 	return
 
+/obj/structure/machinery/door/airlock/dropship_hatch/initialize_pass_flags(var/datum/pass_flags_container/PF)
```
Comment (BeagleGaming1) [inline code/game/machinery/doors/airlock_types.dm:830]:
```suggestion
/obj/structure/machinery/door/airlock/dropship_hatch/initialize_pass_flags(datum/pass_flags_container/PF)
```
You can't do var/ in these places anymore

File: code/game/machinery/doors/multi_tile.dm [hunk 3]
```diff
@@ -275,6 +275,10 @@
 /obj/structure/machinery/door/airlock/multi_tile/almayer/dropshiprear/ex_act(severity)
 	return
 
+/obj/structure/machinery/door/airlock/multi_tile/almayer/dropshiprear/initialize_pass_flags(var/datum/pass_flags_container/PF)
```
Comment (BeagleGaming1) [inline code/game/machinery/doors/multi_tile.dm:278]:
```suggestion
/obj/structure/machinery/door/airlock/multi_tile/almayer/dropshiprear/initialize_pass_flags(datum/pass_flags_container/PF)
```

File: code/game/machinery/doors/poddoor.dm [hunk 4]
```diff
@@ -300,6 +300,11 @@
 	unslashable = TRUE
 	unacidable = TRUE
 
+/obj/structure/machinery/door/poddoor/shutters/almayer/locked/initialize_pass_flags(var/datum/pass_flags_container/PF)
```
Comment (BeagleGaming1) [inline code/game/machinery/doors/poddoor.dm:303]:
```suggestion
/obj/structure/machinery/door/poddoor/shutters/almayer/locked/initialize_pass_flags(datum/pass_flags_container/PF)
```

File: code/game/machinery/doors/shutters.dm [hunk 5]
```diff
@@ -131,6 +131,11 @@
 /obj/structure/machinery/door/poddoor/shutters/almayer/pressure/ex_act(severity)
 	return
 
+/obj/structure/machinery/door/poddoor/shutters/almayer/pressure/initialize_pass_flags(var/datum/pass_flags_container/PF)
```
Comment (BeagleGaming1) [inline code/game/machinery/doors/shutters.dm:134]:
```suggestion
/obj/structure/machinery/door/poddoor/shutters/almayer/pressure/initialize_pass_flags(datum/pass_flags_container/PF)
```

File: code/game/objects/structures.dm [hunk 6]
```diff
@@ -20,6 +20,11 @@
 	if(climbable)
 		verbs += /obj/structure/proc/climb_on
 
+/obj/structure/initialize_pass_flags(var/datum/pass_flags_container/PF)
```
Comment (BeagleGaming1) [inline code/game/objects/structures.dm:23]:
```suggestion
/obj/structure/initialize_pass_flags(datum/pass_flags_container/PF)
```

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 7]
```diff
@@ -54,27 +94,47 @@
 	if(burrow)
 		addtimer(CALLBACK(src, PROC_REF(process_burrow)), 1 SECONDS)
 
+/mob/living/carbon/xenomorph/proc/process_burrow_impaler(var/turf/T = get_turf(src))
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:97]:
```suggestion
/mob/living/carbon/xenomorph/proc/process_burrow_impaler(turf/T = get_turf(src))
```

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 8]
```diff
@@ -214,3 +274,313 @@
...
+	X.visible_message(SPAN_XENODANGER("[X] shoots spikes though the ground in front of it!"), SPAN_XENODANGER("You shoot your spikes though the ground in front of you!"))
+
+	// Loop through our turfs, finding any humans there and dealing damage to them
+	INVOKE_ASYNC(src, PROC_REF(handle_damage), X, target_turfs, telegraph_atom_list, damage)
+
+	apply_cooldown()
+	..()
+	return
+
+/datum/action/xeno_action/activable/burrowed_spikes/proc/handle_damage(var/mob/living/carbon/xenomorph/X, target_turfs, telegraph_atom_list, damage)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:349]:
```suggestion
/datum/action/xeno_action/activable/burrowed_spikes/proc/handle_damage(mob/living/carbon/xenomorph/X, target_turfs, telegraph_atom_list, damage)
```

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 9]
```diff
@@ -214,3 +274,313 @@
...
+	..()
+	return
+
+/datum/action/xeno_action/activable/sunken_tail/can_use_action()
+	var/mob/living/carbon/xenomorph/X = owner
+	if(!X) return
+	if(X.burrow) return FALSE
+	return ..()
+
+/datum/action/xeno_action/activable/sunken_tail/proc/handle_damage(var/mob/living/carbon/xenomorph/X, target_turfs, damage, var/spike_circle = FALSE)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:431]:
```suggestion
/datum/action/xeno_action/activable/sunken_tail/proc/handle_damage(mob/living/carbon/xenomorph/X, target_turfs, damage, var/spike_circle = FALSE)
```

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 10]
```diff
@@ -214,3 +274,313 @@
...
+	..()
+	return
+
+/datum/action/xeno_action/onclick/ensconce/can_use_action()
+	var/mob/living/carbon/xenomorph/X = owner
+	if(!X) return
+	if(X.burrow) return FALSE
+	return ..()
+
+/datum/action/xeno_action/onclick/ensconce/proc/fortify_switch(var/mob/living/carbon/xenomorph/X, var/fortify_state)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:537]:
```suggestion
/datum/action/xeno_action/onclick/ensconce/proc/fortify_switch(mob/living/carbon/xenomorph/X, var/fortify_state)
```

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 11]
```diff
@@ -214,3 +274,313 @@
...
+		X.update_canmove()
+		X.fortify = FALSE
+
+/datum/action/xeno_action/onclick/ensconce/proc/death_check()
+	SIGNAL_HANDLER
+
+	UnregisterSignal(owner, COMSIG_MOB_DEATH)
+	fortify_switch(owner, FALSE)
+
+/datum/action/xeno_action/onclick/ensconce/proc/process_ensconce(var/mob/living/carbon/xenomorph/xeno)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:577]:
```suggestion
/datum/action/xeno_action/onclick/ensconce/proc/process_ensconce(mob/living/carbon/xenomorph/xeno)
```

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 12]
```diff
@@ -214,3 +274,313 @@
...
+	..()
+	return
+
+/datum/action/xeno_action/activable/sunken_tail/can_use_action()
+	var/mob/living/carbon/xenomorph/X = owner
+	if(!X) return
+	if(X.burrow) return FALSE
+	return ..()
+
+/datum/action/xeno_action/activable/sunken_tail/proc/handle_damage(mob/living/carbon/xenomorph/X, target_turfs, damage, var/spike_circle = FALSE)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:431]:
```suggestion
/datum/action/xeno_action/activable/sunken_tail/proc/handle_damage(mob/living/carbon/xenomorph/X, target_turfs, damage, spike_circle = FALSE)
```
Guess I missed a few

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 13]
```diff
@@ -214,3 +274,313 @@
...
+	..()
+	return
+
+/datum/action/xeno_action/onclick/ensconce/can_use_action()
+	var/mob/living/carbon/xenomorph/X = owner
+	if(!X) return
+	if(X.burrow) return FALSE
+	return ..()
+
+/datum/action/xeno_action/onclick/ensconce/proc/fortify_switch(mob/living/carbon/xenomorph/X, var/fortify_state)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:537]:
```suggestion
/datum/action/xeno_action/onclick/ensconce/proc/fortify_switch(mob/living/carbon/xenomorph/X, fortify_state)
```

File: colonialmarines.dme [hunk 14]
```diff
@@ -1893,6 +1893,7 @@
 #include "code\modules\mob\living\carbon\xenomorph\mutators\strains\ravager\berserker.dm"
 #include "code\modules\mob\living\carbon\xenomorph\mutators\strains\ravager\hedgehog.dm"
 #include "code\modules\mob\living\carbon\xenomorph\mutators\strains\runner\acid.dm"
+#include "code\modules\mob\living\carbon\xenomorph\mutators\strains\burrower\impaler.dm"
```
Comment (BeagleGaming1) [inline colonialmarines.dme:1896]: Linter also gets angry if this isn't in alphabetical order

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_abilities.dm [hunk 15]
```diff
@@ -11,7 +11,7 @@
 
 /datum/action/xeno_action/activable/burrow/use_ability(atom/A)
 	var/mob/living/carbon/xenomorph/X = owner
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_abilities.dm:13]: single letter var

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_abilities.dm [hunk 16]
```diff
@@ -11,7 +11,7 @@
 
 /datum/action/xeno_action/activable/burrow/use_ability(atom/A)
```
Comment (BeagleGaming1) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_abilities.dm:12]: single letter var

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 17]
```diff
@@ -214,3 +274,313 @@
...
+
+	playsound(X.loc, 'sound/effects/burrower_attack.ogg', 40)
+	X.visible_message(SPAN_XENODANGER("[X] shoots spikes though the ground in front of it!"), SPAN_XENODANGER("You shoot your spikes though the ground in front of you!"))
+
+	// Loop through our turfs, finding any humans there and dealing damage to them
+	INVOKE_ASYNC(src, PROC_REF(handle_damage), X, target_turfs, telegraph_atom_list, damage)
+
+	apply_cooldown()
+	..()
+	return
```
[the remark this answers is not in the document]
Comment (Unknownity) [reply code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:347]: scroll up

File: code/modules/mob/living/living_verbs.dm [hunk 18]
```diff
@@ -11,6 +11,12 @@
 		to_chat(src, SPAN_WARNING("You can't resist in your current state."))
 		return
 
+	if(isxeno(src))
+		var/mob/living/carbon/xenomorph/Xeno = src
```
[the remark this answers is not in the document]
Comment (Unknownity) [reply code/modules/mob/living/living_verbs.dm:15]: meant to not be a snake case iirc (is_xeno and isxeno are different afaik) correct me if wrong

File: code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm [hunk 19]
```diff
@@ -214,3 +275,313 @@
...
+		if(xeno.selected_ability != src)
+			button.icon_state = "template"
+
+	..()
+	return
+
+/datum/action/xeno_action/onclick/ensconce/can_use_action()
+	var/mob/living/carbon/xenomorph/xeno = owner
+	if(!xeno) return
+	if(xeno.burrow) return FALSE
```
Comment (harryob) [inline code/modules/mob/living/carbon/xenomorph/abilities/burrower/burrower_powers.dm:535]:
```suggestion
	if(!xeno) 
		return
	if(xeno.burrow) 
		return FALSE
```

**Review by harryob — CHANGES_REQUESTED** [review]

**Comment by Venuska1117** [comment]
Yo, after playing it i found some issues, one of it is...
membranes block your range attack if you are in forify mode.

**Comment by Unknownity** [comment]
> Yo, after playing it i found some issues, one of it is... membranes block your range attack if you are in forify mode.
> 
> _and i really want to be able use single stab skill when burrowed, aww..._

intended, spikes can only go through window frames/windows and airlocks.
also tail stab not working on regular burrow is also intended based on a previous balance change, I only added it back for Impaler as they aren't entirely underground and provides risk due to being in melee weed removal range (aka they have counterplay)

**Comment by Venuska1117** [comment]
> > Yo, after playing it i found some issues, one of it is... membranes block your range attack if you are in forify mode.
> > _and i really want to be able use single stab skill when burrowed, aww..._
> 
> intended, spikes can only go through window frames/windows and airlocks. also tail stab not working on regular burrow is also intended based on a previous balance change, I only added it back for Impaler as they aren't entirely underground and provides risk due to being in melee weed removal range (aka they have counterplay)

yes but the ground tail stab, when you are foryfied don't work if you have membrane in front of you, it think its wall.

**Review by morrowwolf — COMMENTED** [review]
Heya!

I got some questions for ya I'd like you to fill out:

Gameplay:
Does it feel good to use?
Does it feel good to play against?
Does it affect other systems?

Roleplay:
Does it increase immersion?
Does it increase communication/interaction between players?

Simulation:
Does it fit in the CM universe aesthetically?
Does it fit in the lore of CM?

**Comment by Unknownity** [comment]
> Heya!
> 
> I got some questions for ya I'd like you to fill out:
> 
> Gameplay: Does it feel good to use? Does it feel good to play against? Does it affect other systems?
> **Its attacks are telegraphed and readable, have time to react against and are only strong when the strain uses the ability which locks them in place which makes them vulnerable to attacks, there are two ways of counterplay against the strain, one is more of a skill equals reward which is sprite click shooting the borrower to damage it, the other is removing the weeds to force it out. Playing against is equal to a boiler, it is annoying but not impossible. as time goes on, players will be able to adapt to the meta against the strain. The strain has been nerfed several times up until this iteration. The only existing issue regarding it is the way how fire affects weeds, it does not instantly remove them which means the borrower has a grace period of being able to move across the flamed tiles before the weeds go away and not be forcibly surfaced, so the only way of instantly flushing them out is through meleeing the tiles, spraying chemicals containing a specific property that removes weeds and explosions. I do not want to change how fire and small explosions affect weeds as that is a massive balance change but If I'm gonna throw my opinion on this is that yes, fire should instantly remove weeds. (except green)**
> Roleplay: Does it increase immersion? Does it increase communication/interaction between players?
> **The strain is more immersive than the original that's for sure, as for the second question I am unable to answer in full confidence due to a lack of observational data, in theory, it should be like a boiler-lite in terms of communication/interaction between the xenomorphs.**
> Simulation: Does it fit in the CM universe aesthetically? Does it fit in the lore of CM?
> **Since Burrowers do not actually exist in the Aliens media, all we have is CM's iteration of what a Burrower is. the original strain of the Burrower is a combat support caste, primarily intended to assist xenomorphs fast traveling around the map with tunnels, fortifying key locations and assisting backliners in general, it is able to solo but not as effective as a lurker against more than 1-2 marines. The Impaler strain is the opposite, instead of combat supporting backliners, it is now a frontline combat support caste (while the original could assist in frontlines, it's usually a waste of their potential). Its advantage is helping in defending the xeno side of the frontline if marines push by slowing them down through its telegraphed attacks and when xenos start advancing, it provides additional pressure to the marines by acting as sort of a mobile environmental hazard.
While the strain is a direct inspiration from the Starcraft lurker, it wouldn't be too far off a xenomorph considering one of their main weapons are their tails, this one simply specialized to ONLY use its tail as its weapon of choice. It only being able to burrow while on weeds is merely a gameplay limitation for balance.**
> Bear in mind, these don't all need to be yes. Just curious what your take on it is.

**Comment by ihatethisengine** [comment]
I think another way to solve flame vs. weeds issue is just to make burrower vulnerable to fire even if burrowed. So impaler would be forced to disengage, surface and resist. This will create an easy and effective way to fight impalers and probably reduce the amount of salt.

**Comment by Venuska1117** [comment]
> I think another way to solve flame vs. weeds issue is just to make impaler vulnerable to fire even if burrowed. So impaler would be forced to disengage, surface and resist. This will create an easy and effective way to fight impalers and probably reduce the amount of salt.

its like said to burn whole fish in water if only their fin stick out of water.|

Just make fire deal damage to them if they on burning tiles but dont ignite burrower (with make more sense)

**Review by sleepynecrons — APPROVED** [review]
sprites approved

**Review by morrowwolf — CHANGES_REQUESTED** [review]
Alright, this is not gonna be what anyone wants to hear: I'm not sure I like the strain.

I feel it is frustrating to fight.
I feel adding more ranged options for xenos is the opposite of the direction we should be going.
I feel the lurker sounds and inspiration are a little too on the nose for our setting.

Some of this I can overlook. The one thing I can't overlook is another ranged strain for xenos. I'm happy to see spikes or moving around underground as it is but it has to be much closer range.

If you can overcome those requirements we can take a look again. Sadly, this is going to require pretty much an entire rework.

**Comment by Venuska1117** [comment]
> Alright, this is not gonna be what anyone wants to hear: I'm not sure I like the strain.
> 
> I feel it is frustrating to fight. I feel adding more ranged options for xenos is the opposite of the direction we should be going. I feel the lurker sounds and inspiration are a little too on the nose for our setting.
> 
> Some of this I can overlook. The one thing I can't overlook is another ranged strain for xenos. I'm happy to see spikes or moving around underground as it is but it has to be much closer range.
> 
> If you can overcome those requirements we can take a look again. Sadly, this is going to require pretty much an entire rework.

this strain was hard to play when it was TM, because marines used granades, flamethrower, and they just obligerated you if you were not careful enough, yes it have ranged skill (the impale), but it make you immobile in one spot, and you cant be behind objects, with make you vonurelable, when you walk it just create 1x1 with is not effective as it sound, when you move while burrowed, spikes wave don't go trough cades, its only good when marines are out of cades, but because of slow attack and attack zone indicator most of time they just avoid it, not to mention if you somehow get out of weeds you are 80% times dead, i saw like 3 impaler burrowers die in like 10min into fight because of nades/mortars/clusterOB (anything that make them go out weeds)

But i can agree that its hard to hit when burrowed it can get increased hit chance when burrower is closer to marines gunfire (more easy to hit)

**Comment by Comxy** [comment]
NOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOOO

**Comment by IsDogHour** [comment]
The best looking addition to the xeno side of CM in a long time gone. How depressing.

**Comment by MPhonks** [comment]
Actually sad.

**Pull request closed without merging**